src/rgw/driver/d4n/d4n_policy.h

```diff
@@ -1,104 +1,150 @@
-#ifndef CEPH_D4NPOLICY_H
-#define CEPH_D4NPOLICY_H
+#pragma once
 
 #include <string>
 #include <iostream>
-#include <cpp_redis/cpp_redis>
 #include "rgw_common.h"
 #include "d4n_directory.h"
 #include "../../rgw_redis_driver.h"
 
+#define dout_subsys ceph_subsys_rgw
+#define dout_context g_ceph_context
+
 namespace rgw { namespace d4n {
 
 class CachePolicy {
   private:
-    cpp_redis::client client;
-    Address addr;
+    struct Entry : public boost::intrusive::list_base_hook<> {
+      std::string key;
+      uint64_t offset;
+      uint64_t len;
+      std::string version;
+      Entry(std::string& key, uint64_t offset, uint64_t len, std::string version) : key(key), offset(offset), 
+                                                                                     len(len), version(version) {}
+    };
+    
+    //The disposer object function
+    struct Entry_delete_disposer {
+      void operator()(Entry *e) {
+        delete e;
+      }
+    };
+    typedef boost::intrusive::list<Entry> List;
 
   public:
     CephContext* cct;
 
-    CachePolicy() : addr() {}
-    virtual ~CachePolicy() = default;
+    CachePolicy() {}
+    virtual ~CachePolicy() = default; 
 
-    virtual void init(CephContext *_cct) {
-      cct = _cct;
-      addr.host = cct->_conf->rgw_d4n_host;
-      addr.port = cct->_conf->rgw_d4n_port;
+    virtual int init(CephContext *cct, const DoutPrefixProvider* dpp) {
+      this->cct = cct;
+      return 0;
     }
-    virtual int find_client(const DoutPrefixProvider* dpp, cpp_redis::client* client) = 0;
-    virtual int exist_key(std::string key) = 0;
-    virtual Address get_addr() { return addr; }
-    virtual int get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheNode) = 0;
-    virtual uint64_t eviction(const DoutPrefixProvider* dpp, rgw::cache::CacheDriver* cacheNode) = 0;
-    virtual void insert(const DoutPrefixProvider* dpp, std::string& key, uint64_t offset, uint64_t len, rgw::cache::CacheDriver* cacheNode) = 0;
+    virtual int exist_key(std::string key, optional_yield y) = 0;
+    virtual int get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheNode, optional_yield y) = 0;
+    virtual uint64_t eviction(const DoutPrefixProvider* dpp, rgw::cache::CacheDriver* cacheNode, optional_yield y) = 0;
+    virtual void insert(const DoutPrefixProvider* dpp, std::string& key, uint64_t offset, uint64_t len, std::string version, rgw::cache::CacheDriver* cacheNode, optional_yield y) = 0;
+    virtual bool erase(const DoutPrefixProvider* dpp, const std::string& key) = 0;
+    virtual void shutdown() = 0;
+
+    friend class LFUDAPolicy;
```

Comment:
Why do you have these as `friend` classes when they inherit from CachePolicy? Is it to access `Entry`? You can make `Entry` as a `protected` and access them in LRU and LFUDA Policy driver classes.